File: 백준/Gold/1753. 최단경로/최단경로.py
import sys
import heapq

V, E = map(int,sys.stdin.readline().split())
start = int(sys.stdin.readline())
adj = [[] for _ in range(V+1)]
dist = [float('inf')] * (V+1)
heap = []
for e in range(E) :
    u, v, w = map(int,sys.stdin.readline().split())
    adj[u].append((w,v))

def dijkstra(start) :
    dist[start] = 0
    heapq.heappush(heap,(0,start))

    while heap :
        now_w, now = heapq.heappop(heap)

        if dist[now] < now_w :
            continue

        for w, next in adj[now] :
            next_w = w + now_w
            if next_w < dist[next] :
                dist[next] = next_w
                heapq.heappush(heap,(next_w, next))
dijkstra(start)

for i in range(1,V+1) :
    if dist[i] == float('inf') :
        print('INF')
    else :
        print(dist[i])

File: 백준/Gold/1753. 최단경로/최단경로.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

class Main {
    static int V, E;
    static int start;
    static ArrayList<int[]>[] adj;
    static int[] dist;
    static PriorityQueue<int[]> heap;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        V = Integer.parseInt(st.nextToken());
        E = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        start = Integer.parseInt(st.nextToken());

        adj = new ArrayList[V + 1];
        for (int i = 1; i <= V; i++) {
            adj[i] = new ArrayList<>();
        }

        for (int e = 0; e < E; e++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());
            adj[u].add(new int[]{w, v});
        }

        dist = new int[V + 1];
        heap = new PriorityQueue<>((a, b) -> Integer.compare(a[0], b[0]));

        dijkstra(start);

        for (int i = 1; i <= V; i++) {
            if (dist[i] == Integer.MAX_VALUE) {
                System.out.println("INF");
            } else {
                System.out.println(dist[i]);
            }
        }
    }

    static void dijkstra(int start) {
        for (int i = 1; i <= V; i++) {
            dist[i] = Integer.MAX_VALUE;
        }

        dist[start] = 0;
        heap.offer(new int[]{0, start});

        while (!heap.isEmpty()) {
            int[] current = heap.poll();
            int nowWeight = current[0];
            int now = current[1];

            if (dist[now] < nowWeight) {
                continue;
            }

            for (int[] edge : adj[now]) {
                int nextWeight = edge[0] + nowWeight;
                int next = edge[1];
                if (nextWeight < dist[next]) {
                    dist[next] = nextWeight;
                    heap.offer(new int[]{nextWeight, next});
                }
            }
        }
    }
}